### ISSUE TITLE
Multiple issues on GCE with internal load balancer and no DNS

### ISSUE THREAD

**Issue by kkoppel** [kubernetes/kops#15226]
/kind bug

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**
Client version: 1.26.2 (git-v1.26.2)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**
1.26.2

**3. What cloud provider are you using?**
GCE

**4. What commands did you run?  What is the simplest way to reproduce this issue?**
```
$ kops create cluster test.k8s \
  --admin-access=10.164.0.0/20 \
  --api-loadbalancer-type=internal \
  --cloud=gce \
  --dns=none \
  --network-id=default \
  --subnets=default \
  --topology=private \
  --set spec.api.additionalSANs=10.164.0.10 \
  --ssh-access=10.164.0.0/20,35.235.240.0/20 \
  --state=gs://mybucket \
  --project=myproject \
  --zones=europe-west4-a
$ kops get all --name=test.k8s --state=gs://mybucket -oyaml > test.yaml
```

Edited the test.yaml file to set a specific internal IP for the LB. This seems to be necessary to get kOps to set up any forwarding rules at all. Then:
```
$ kops replace -f test.yaml --state=gs://mybucket
$ kops update cluster --name=test.k8s --yes --admin --state=gs://mybucket
```

**5. What happened after the commands executed?**
kubeconfig export fails at the end of the `kops update cluster` command:
```
I0313 15:28:25.783665  896113 backend_service.go:127] GCE creating backend service: "api-test-k8s"
I0313 15:28:38.398918  896113 executor.go:111] Tasks: 65 done / 67 total; 2 can run
I0313 15:28:58.540009  896113 executor.go:111] Tasks: 67 done / 67 total; 0 can run
I0313 15:28:58.663408  896113 update_cluster.go:323] Exporting kubeconfig for cluster
Error: error getting ingress status: error getting ForwardingRule "api-test-k8s": googleapi: Error 404: The resource 'projects/myproject/regions/europe-west4/forwardingRules/api-test-k8s' was not found, notFound
```

With this cluster configuration kOps creates 2 ForwardingRule resources as follows:
```
  ForwardingRule/default-test-k8s
        Ports                   [443]
        RuleIPAddress           10.164.0.10
        IPProtocol              TCP
        LoadBalancingScheme     INTERNAL
        Network                 name:default id:default
        Subnetwork              name:default id:default
        BackendService          name:api-test-k8s id:api-test-k8s

  ForwardingRule/kops-controller-default-test-k8s
        Ports                   [3988]
        RuleIPAddress           10.164.0.10
        IPProtocol              TCP
        LoadBalancingScheme     INTERNAL
        Network                 name:default id:default
        Subnetwork              name:default id:default
        BackendService          name:api-test-k8s id:api-test-k8s
```

As a workaround I can change the 1st ForwardingRule name manually and then successfully export kubeconfig like this:
```
$ gcloud compute forwarding-rules export default-test-k8s --region=europe-west4 --destination=kops_fwd_rule.yaml
$ gcloud compute forwarding-rules delete default-test-k8s --region=europe-west4
$ sed -i 's/default-test-k8s/api-test-k8s/g' kops_fwd_rule.yaml
$ gcloud compute forwarding-rules import api-test-k8s --region=europe-west4 --source=kops_fwd_rule.yaml
$ kops export kubeconfig --admin --name test.k8s --state=gs://mybucket
kOps has set your kubectl context to test.k8s
```

I can then try validating the cluster, but the worker node never joins the cluster:
```
$ kops validate cluster --state=gs://mybucket
Using cluster from kubectl context: test.k8s

Validating cluster test.k8s

I0313 15:41:21.669443  896200 gce_cloud.go:295] Scanning zones: [europe-west4-a europe-west4-b europe-west4-c]
INSTANCE GROUPS
NAME                            ROLE            MACHINETYPE     MIN     MAX     SUBNETS
control-plane-europe-west4-a    ControlPlane    e2-medium       1       1       europe-west4
nodes-europe-west4-a            Node            e2-medium       1       1       europe-west4

NODE STATUS
NAME                                    ROLE            READY
control-plane-europe-west4-a-ldh2       control-plane   True

VALIDATION ERRORS
KIND    NAME                                                                                                                            MESSAGE
Machine https://www.googleapis.com/compute/v1/projects/myproject/zones/europe-west4-a/instances/nodes-europe-west4-a-xnm2       machine "https://www.googleapis.com/compute/v1/projects/myproject/zones/europe-west4-a/instances/nodes-europe-west4-a-xnm2" has not yet joined cluster
Pod     kube-system/coredns-5779545d4-tm5vj                                                                                             system-cluster-critical pod "coredns-5779545d4-tm5vj" is pending
Pod     kube-system/coredns-autoscaler-7f4ddcc6c7-8pcbl                                                                                 system-cluster-critical pod "coredns-autoscaler-7f4ddcc6c7-8pcbl" is pending

Validation Failed
Error: validation failed: cluster not yet healthy
```

**6. What did you expect to happen?**
I expected `kops update cluster` to finish successfully and after some time the cluster should validate successfully as well.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2023-03-13T15:25:28Z"
  generation: 1
  name: test.k8s
spec:
  additionalSans:
  - 10.164.0.10
  api:
    loadBalancer:
      subnets:
      - name: default
        privateIPv4Address: 10.164.0.10
      type: Internal
  authorization:
    rbac: {}
  channel: stable
  cloudConfig: {}
  cloudProvider: gce
  configBase: gs://mybucket/test.k8s
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: control-plane-europe-west4-a
      name: a
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: control-plane-europe-west4-a
      name: a
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeProxy:
    enabled: false
  kubelet:
    anonymousAuth: false
  kubernetesApiAccess:
  - 10.164.0.0/20
  kubernetesVersion: 1.26.2
  networkID: default
  networking:
    cilium:
      enableNodePort: true
  nonMasqueradeCIDR: 100.64.0.0/10
  project: myproject
  sshAccess:
  - 10.164.0.0/20
  - 35.235.240.0/20
  subnets:
  - id: default
    name: europe-west4
    region: europe-west4
    type: Private
  topology:
    dns:
      type: None
    masters: private
    nodes: private

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-13T15:25:29Z"
  labels:
    kops.k8s.io/cluster: test.k8s
  name: control-plane-europe-west4-a
spec:
  image: ubuntu-os-cloud/ubuntu-2004-focal-v20230302
  machineType: e2-medium
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - europe-west4
  zones:
  - europe-west4-a

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-03-13T15:25:29Z"
  labels:
    kops.k8s.io/cluster: test.k8s
  name: nodes-europe-west4-a
spec:
  image: ubuntu-os-cloud/ubuntu-2004-focal-v20230302
  machineType: e2-medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    cloud.google.com/metadata-proxy-ready: "true"
  role: Node
  subnets:
  - europe-west4
  zones:
  - europe-west4-a
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

Verbose log for the kubeconfig export:
```
$ kops export kubeconfig --admin --name test.k8s --state=gs://mybucket -v 10
I0313 13:45:08.107695  895441 factory.go:83] state store gs://mybucket
I0313 13:45:08.130582  895441 gsfs.go:274] Reading file "gs://mybucket/test.k8s/config"
I0313 13:45:08.200028  895441 clientset.go:103] Using keystore path: "gs://mybucket/test.k8s/pki"
I0313 13:45:08.228745  895441 gce_cloud.go:168] running with GCE credentials: email=, scope=https://www.googleapis.com/auth/cloud-platform
I0313 13:45:08.228895  895441 gce_cloud.go:309] Querying GCE to find ForwardingRules for API ("api-test-k8s")
Error: error getting ingress status: error getting ForwardingRule "api-test-k8s": googleapi: Error 404: The resource 'projects/myproject/regions/europe-west4/forwardingRules/api-test-k8s' was not found, notFound
```

I can see the following lines repeating in the worker node journal:
```
Mar 13 15:48:55 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:55.066339    3179 kubelet_node_status.go:352] "Setting node annotation to enable volume controller attach/detach"
Mar 13 15:48:55 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:55.072230    3179 kubelet_node_status.go:669] "Recording event message for node" node="nodes-europe-west4-a-xnm2" event="NodeHasSufficientMemory"
Mar 13 15:48:55 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:55.072546    3179 kubelet_node_status.go:669] "Recording event message for node" node="nodes-europe-west4-a-xnm2" event="NodeHasNoDiskPressure"
Mar 13 15:48:55 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:55.072724    3179 kubelet_node_status.go:669] "Recording event message for node" node="nodes-europe-west4-a-xnm2" event="NodeHasSufficientPID"
Mar 13 15:48:55 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:55.072887    3179 kubelet_node_status.go:70] "Attempting to register node" node="nodes-europe-west4-a-xnm2"
Mar 13 15:48:55 nodes-europe-west4-a-xnm2 kubelet[3179]: E0313 15:48:55.082674    3179 kubelet_node_status.go:92] "Unable to register node with API server" err="Post \"https://api.internal.test.k8s/api/v1/nodes\": dial tcp: lookup api.internal.test.k8s: no such host" node="nodes-europe-west4-a-xnm2"
Mar 13 15:48:56 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:56.010878    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:48:57 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:57.013369    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:48:58 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:58.019428    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:48:59 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:48:59.011836    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:49:00 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:49:00.013842    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:49:01 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:49:01.012348    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:49:01 nodes-europe-west4-a-xnm2 kubelet[3179]: E0313 15:49:01.310368    3179 controller.go:146] failed to ensure lease exists, will retry in 7s, error: Get "https://api.internal.test.k8s/apis/coordination.k8s.io/v1/namespaces/kube-node-lease/leases/nodes-europe-west4-a-xnm2?timeout=10s": dial tcp: lookup api.internal.test.k8s: no such host
Mar 13 15:49:02 nodes-europe-west4-a-xnm2 kubelet[3179]: I0313 15:49:02.010420    3179 csi_plugin.go:913] Failed to contact API server when waiting for CSINode publishing: Get "https://api.internal.test.k8s/apis/storage.k8s.io/v1/csinodes/nodes-europe-west4-a-xnm2": dial tcp: lookup api.internal.test.k8s: no such host
```

**9. Anything else do we need to know?**
Manually adding a record for api.internal.test.k8s to /etc/hosts with the LB IP fixes the worker node and after a small delay cluster validation succeeds.

**Comment by rifelpet** [comment]
For reference, it looks like for public API load balancers we use the `api` name:

https://github.com/kubernetes/kops/blob/e63cb44bd06e4b5d079f53f164d526035d29f84e/pkg/model/gcemodel/api_loadbalancer.go#L70-L71

but internal API load balancers use the subnet name:

https://github.com/kubernetes/kops/blob/e63cb44bd06e4b5d079f53f164d526035d29f84e/pkg/model/gcemodel/api_loadbalancer.go#L180-L181

but the logic used for kubeconfig generation always assumes `api`:

https://github.com/kubernetes/kops/blob/e63cb44bd06e4b5d079f53f164d526035d29f84e/upup/pkg/fi/cloudup/gce/gce_cloud.go#L309-L324

This logic needs to be updated to recognize internal API load balancers and generate the appropriate forwarding rule name

**Comment by johngmyers** [comment]
Office hours: cherrypick to 1.27

**Comment by hakman** [comment]
This should be fixed for 1.27.0. Thanks for reporting @kkoppel!
